src/v/compression/internal/gzip_compressor.cc

```diff
@@ -117,106 +117,114 @@ class gzip_decompression_codec {
 
 private:
     bool _init{false};
-    const char* _input;
-    size_t _input_size;
+
+    const iobuf& _input;
+    iobuf::const_iterator _input_chunk;
+
     gz_header _hdr; // needed for gzip
     z_stream _stream;
 };
 
 iobuf gzip_compressor::compress(const iobuf& b) {
-    ss::temporary_buffer<char> obuf;
-    {
-        gzip_compression_codec def;
-        def.reset();
-        z_stream& strm = def.stream();
-        /* Calculate maximum compressed size and
-         * allocate an output buffer accordingly, being
-         * prefixed with the Message header. */
-        const size_t output_size = deflateBound(&strm, b.size_bytes());
-        obuf = ss::temporary_buffer<char>(output_size);
+    const size_t max_chunk_size = details::io_allocation_size::max_chunk_size;
 
-        // NOLINTNEXTLINE
-        strm.next_out = (unsigned char*)obuf.get_write();
-        strm.avail_out = output_size;
-
-        /* Iterate through each segment and compress it. */
-        for (auto& io : b) {
-            // zlib is not const correct
-            // NOLINTNEXTLINE
-            strm.next_in = (unsigned char*)io.get();
-            strm.avail_in = io.size();
-            throw_if_zstream_error(
-              "gzip error compressing chunk: {}", deflate(&strm, Z_NO_FLUSH));
+    gzip_compression_codec def;
+    def.reset();
+    z_stream& strm = def.stream();
+
+    const size_t output_size_estimate = deflateBound(&strm, b.size_bytes());
+    size_t output_chunk_size = std::min(max_chunk_size, output_size_estimate);
+
+    ss::temporary_buffer<char> obuf(output_chunk_size);
+    strm.next_out = reinterpret_cast<unsigned char*>(obuf.get_write());
+    strm.avail_out = obuf.size();
+
+    iobuf ret;
+    auto frag_i = b.begin();
+    while (true) {
+        if (strm.avail_in == 0 && frag_i != b.end()) {
+            strm.next_in = const_cast<unsigned char*>(
+              reinterpret_cast<const unsigned char*>(frag_i->get()));
+            strm.avail_in = frag_i->size();
```

Comment:
it'd be very unlikely for it to happen, but iobuf does allow empty fragments (which we had discovered with the fuzzer). ideally everything below shakes out correctly for that case?